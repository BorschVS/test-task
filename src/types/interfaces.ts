import { ReactNode, ElementType, MouseEventHandler } from 'react';

import { SET_FORM_DATA } from '../redux/ducks/form';
import { CHEAP_VALUE, FAST_VALUE } from '../constants';

import {
  GET_FLIGHTS,
  SET_FLIGHTS,
  SET_FLIGHTS_LOADING,
  SET_FLIGHTS_ERROR,
  SET_STOPS_FILTER,
  SET_ALL_STOPS,
  SET_STOPS_FILTER_STATUS,
  SET_CATEGORY_FILTER,
} from '../redux/ducks/flights';
import { SET_CURRENT_FLIGHT, SET_IS_SHOWING } from '../redux/ducks/modal';
import {
  GET_IMAGE_FAILURE,
  GET_IMAGE_REQUEST,
  GET_IMAGE_SUCCESS,
} from '../redux/ducks/images';

// redux

export interface FlightData {
  id: string;
  searchId: string;
  price: number;
  carrier: string;
  segments: FlightSegments[];
}

export interface FlightSegments {
  origin: string;
  destination: string;
  date: string;
  stops: string[];
  duration: number;
}

export interface FlightsState {
  availableFlights: FlightData[];
  filteredFlights: FlightData[];
  stopsFilter: number[];
  stopsFilterStatus: boolean;
  flightCategoryFilter: string;
  loading: boolean;
  error: Error | string | null;
}

export interface ModalState {
  isShowing: boolean;
  flightId: string;
}

export interface FormState {
  formData: string;
}

export interface ImagesState {
  images: { [key: string]: string };
  loading: boolean;
  error: string | null;
}

export interface ReduxState {
  flights: FlightsState;
  modal: ModalState;
  form: FormState;
  images: ImagesState;
}

// /flights

// actions

export interface GetFlightsAction {
  type: typeof GET_FLIGHTS;
}

export interface SetFlightsAction {
  type: typeof SET_FLIGHTS;
  payload: FlightData[];
}

export interface SetFlightsLoadingAction {
  type: typeof SET_FLIGHTS_LOADING;
  payload: boolean;
}

export interface SetFlightsErrorAction {
  type: typeof SET_FLIGHTS_ERROR;
  payload: string;
}

export interface SetStopsFilterAction {
  type: typeof SET_STOPS_FILTER;
  payload: number[];
}

export interface SetAllStopsAction {
  type: typeof SET_ALL_STOPS;
  payload: number[];
}

export interface SetStopsFilterStatusAction {
  type: typeof SET_STOPS_FILTER_STATUS;
  payload: boolean;
}

export interface SetCategoryFilterAction {
  type: typeof SET_CATEGORY_FILTER;
  payload: typeof CHEAP_VALUE | typeof FAST_VALUE;
}

// /form

// actions

export interface SetFormDataAction {
  type: typeof SET_FORM_DATA;
  payload: string;
}

// /modal

export interface SetIsShowing {
  type: typeof SET_IS_SHOWING;
  payload: boolean;
}

export interface SetCurrentFlight {
  type: typeof SET_CURRENT_FLIGHT;
  payload: number;
}

// /images

// actions

export interface GetImageRequestAction {
  type: typeof GET_IMAGE_REQUEST;
  payload: string;
}

export interface GetImageSuccessAction {
  type: typeof GET_IMAGE_SUCCESS;
  payload: {
    imageName: string;
    imageUrl: string;
  };
}

export interface GetImageFailureAction {
  type: typeof GET_IMAGE_FAILURE;
  payload: string;
}

// Props

export interface FlightCardProps {
  flightData: FlightData;
}

export interface ModalProps {
  isShowing: boolean;
  hide: MouseEventHandler<HTMLAnchorElement>;
}

// image

export interface Image {
  imageName: string;
  alt?: string;
  component?: ElementType;
  children?: ReactNode;
  width?: number;
  height?: number;
  margin?: string;
}

// theme

export interface Theme {
  palette: {
    background: {
      default: string;
      paper: string;
    };
    text: {
      primary: string;
      secondary: string;
      disabled: string;
    };
    primary: {
      main: string;
      light: string;
      dark: string;
      contrastText: string;
    };
    secondary: {
      main: string;
      light: string;
      dark: string;
      contrastText: string;
    };
    error: {
      main: string;
    };
  };
  typography: {
    fontFamily: string;
    fontWeightMedium: number;
    fontWeightRegular: number;
    fontWeightBold: number;
  };
  breakpoints: {
    values: {
      xs: number;
      sm: number;
      md: number;
      lg: number;
      xl: number;
    };
  };
}
