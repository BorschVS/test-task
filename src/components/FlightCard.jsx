import { Avatar, Typography } from '@mui/material';

import {
  CardBox,
  FactText,
  GroupBox,
  InfoText,
  TextBox,
} from '../styled/FlightCard.styled';

import S7Airlines from '../images/s7.jpg';

const FlightCard = () => (
  <CardBox>
    <GroupBox>
      <Typography
        fontSize={24}
        fontWeight={(theme) => theme.typography.fontWeightMedium}
        textTransform="uppercase"
        color={(theme) => theme.palette.primary.blue}
      >
        14 000 P
      </Typography>
      <Typography
        display="flex"
        gap={1}
        fontSize={24}
        fontWeight={(theme) => theme.typography.fontWeightBold}
      >
        <Avatar
          component={'span'}
          src={S7Airlines}
          alt="S7 Airlines company logo"
        />
        Airlines
      </Typography>
    </GroupBox>
    <GroupBox>
      <TextBox>
        <InfoText>Mow - HKT</InfoText>
        <FactText>10:45 - 00:50</FactText>
      </TextBox>
      <TextBox>
        <InfoText>в пути</InfoText>
        <FactText>21ч 15м</FactText>
      </TextBox>
      <TextBox>
        <InfoText>2 пересадки</InfoText>
        <FactText>HKG, JNB</FactText>
      </TextBox>
    </GroupBox>
    <GroupBox>
      <TextBox>
        <InfoText>Mow - HKT</InfoText>
        <FactText>10:45 - 00:50</FactText>
      </TextBox>
      <TextBox>
        <InfoText>в пути</InfoText>
        <FactText>13ч 30м</FactText>
      </TextBox>
      <TextBox>
        <InfoText>2 пересадки</InfoText>
        <FactText>HKG</FactText>
      </TextBox>
    </GroupBox>
  </CardBox>
);

export default FlightCard;
