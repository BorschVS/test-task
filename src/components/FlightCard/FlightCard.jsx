import { Avatar, Box, Typography } from "@mui/material";
import {
  CardBox,
  FactText,
  InfoText,
  TextBox,
  Title,
} from "./FlightCard.styled";
import S7Airline from "../../images/s7.jpg";

const FlightCard = () => {
  return (
    <CardBox>
      {/* row 1 */}
      <Box display="flex" justifyContent="space-between">
        <Title>14 000 P</Title>
        <Typography
          display="flex"
          gap={1}
          fontSize={24}
          fontWeight={(theme) => theme.typography.fontWeightSemiBold}
        >
          <Avatar src={S7Airline} alt="The company logo" />
          Airlines
        </Typography>
      </Box>
      {/* row 2 */}
      <Box display="flex" justifyContent="space-between">
        <TextBox>
          <InfoText>Mow - HKT</InfoText>
          <FactText>10:45 - 00:50</FactText>
        </TextBox>
        <TextBox>
          <InfoText>в пути</InfoText>
          <FactText>10:45 - 00:50</FactText>
        </TextBox>
        <TextBox>
          <InfoText>2 пересадки</InfoText>
          <FactText>10:45 - 00:50</FactText>
        </TextBox>
      </Box>
      {/* row 3 */}
      <Box display="flex" justifyContent="space-between">
        <TextBox>
          <InfoText>Mow - HKT</InfoText>
          <FactText>10:45 - 00:50</FactText>
        </TextBox>
        <TextBox>
          <InfoText>в пути</InfoText>
          <FactText>10:45 - 00:50</FactText>
        </TextBox>
        <TextBox>
          <InfoText>2 пересадки</InfoText>
          <FactText>10:45 - 00:50</FactText>
        </TextBox>
      </Box>
    </CardBox>
  );
};

export default FlightCard;
