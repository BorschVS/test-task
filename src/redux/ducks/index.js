export { default as flights } from './flights';
export { default as form } from './form';
export { default as modal } from './modal';
export { default as images } from './images';