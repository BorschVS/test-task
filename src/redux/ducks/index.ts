export { flightsReducer } from './flights';
export { formReducer } from './form';
export { modalReducer } from './modal';
export { imagesReducer } from './images';
