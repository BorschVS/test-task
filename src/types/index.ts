import { CHEAP_VALUE, FAST_VALUE } from '../constants';

import {
  GetFlightsAction,
  SetAllStopsAction,
  SetFlightsAction,
  SetFlightsErrorAction,
  SetFlightsLoadingAction,
  SetStopsFilterAction,
  SetStopsFilterStatusAction,
  SetCategoryFilterAction,
  SetFormDataAction,
  SetIsShowing,
  SetCurrentFlight,
  GetImageRequestAction,
  GetImageSuccessAction,
  GetImageFailureAction,
} from 'interfaces';

// /flights

export type FlightActions =
  | GetFlightsAction
  | SetFlightsAction
  | SetFlightsLoadingAction
  | SetFlightsErrorAction
  | SetStopsFilterAction
  | SetAllStopsAction
  | SetStopsFilterStatusAction
  | SetCategoryFilterAction;

export type FlightCategory = typeof CHEAP_VALUE | typeof FAST_VALUE;

// /form

export type FormActions = SetFormDataAction;

// /modal

export type ModalActions = SetIsShowing | SetCurrentFlight;

// /images

export type ImagesActions =
  | GetImageRequestAction
  | GetImageSuccessAction
  | GetImageFailureAction;
