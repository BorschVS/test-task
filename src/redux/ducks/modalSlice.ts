import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { ModalState } from 'types/interfaces';

export const SET_CURRENT_FLIGHT = 'SET_CURRENT_FLIGHT';
export const SET_IS_SHOWING = 'SET_IS_SHOWING';

const initialState: ModalState = {
  isShowing: false,
  flightId: '',
};

const modalSlice = createSlice({
  name: 'modal',
  initialState,
  reducers: {
    setIsShowing(state, action: PayloadAction<boolean>) {
      state.isShowing = action.payload;
    },
    setCurrentFlight(state, action: PayloadAction<string>) {
      state.flightId = action.payload.toString();
    },
  },
});

export const { setIsShowing, setCurrentFlight } = modalSlice.actions;

export default modalSlice.reducer;
