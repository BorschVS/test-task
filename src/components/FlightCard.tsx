import { FC } from 'react';
import { useDispatch } from 'react-redux';
import { Typography } from '@mui/material';

import { AvatarImage } from 'components';

import { useModal } from 'hooks/useModal';

import { FlightCardProps } from 'types/interfaces';

import { setCurrentFlight } from '../redux/ducks/modalSlice';
import { AppDispatch } from 'redux/configureStore';

import {
  formatNumber,
  formatTimeRange,
  formatFlightDuration,
  formatTransfersWordEnding,
} from '../utils';

import {
  CardBox,
  FactText,
  GroupBox,
  InfoText,
  TextBox,
} from 'styled/FlightCard.styled';

const FlightCard: FC<FlightCardProps> = ({ flightData }) => {
  const dispatch: AppDispatch = useDispatch();

  const { isShowing, toggleModal } = useModal();

  const handleCurrentFlight = () => {
    dispatch(setCurrentFlight(flightData.id));
  };

  const handleModal = () => {
    if (isShowing) return;

    toggleModal();
    handleCurrentFlight();
  };

  const outboundSegmentData = flightData.segments[0];
  const returnSegmentData = flightData.segments[1];

  const outboundOrigin = outboundSegmentData.origin;
  const returnOrigin = returnSegmentData.origin;

  const outboundDestination = outboundSegmentData.destination;
  const returnDestination = returnSegmentData.destination;

  const outboundFlightTime = formatTimeRange(
    outboundSegmentData.date,
    outboundSegmentData.duration
  );

  const returnFlightTime = formatTimeRange(
    returnSegmentData.date,
    returnSegmentData.duration
  );

  const outboundFlightDuration = formatFlightDuration(
    outboundSegmentData.date,
    outboundSegmentData.duration
  );

  const returnFlightDuration = formatFlightDuration(
    returnSegmentData.date,
    returnSegmentData.duration
  );

  const outboundFlightTransfers = outboundSegmentData.stops.join(', ');
  const returnFlightTransfers = returnSegmentData.stops.join(', ');

  return (
    <CardBox
      onClick={handleModal}
      isShowing={isShowing}
      data-testid="flight-card"
    >
      <GroupBox>
        <Typography
          fontSize={24}
          fontWeight={(theme) => theme.typography.fontWeightMedium}
          textTransform="uppercase"
          color={(theme) => theme.palette.primary.contrastText}
        >
          {formatNumber(flightData.price)} Р
        </Typography>
        <Typography
          display="flex"
          alignItems={'center'}
          gap={1}
          fontSize={24}
          fontWeight={(theme) => theme.typography.fontWeightBold}
        >
          <AvatarImage
            component="span"
            imageName="s7.jpg"
            alt="S7 Airlines company logo"
          />
          Airlines
        </Typography>
      </GroupBox>
      <GroupBox>
        <TextBox>
          <InfoText>{`${outboundOrigin} - ${outboundDestination}`}</InfoText>
          <FactText>{outboundFlightTime}</FactText>
        </TextBox>
        <TextBox>
          <InfoText>в пути</InfoText>
          <FactText>{outboundFlightDuration}</FactText>
        </TextBox>
        <TextBox>
          <InfoText>
            {formatTransfersWordEnding(outboundSegmentData.stops.length)}
          </InfoText>
          <FactText>{outboundFlightTransfers}</FactText>
        </TextBox>
      </GroupBox>
      <GroupBox>
        <TextBox>
          <InfoText>{`${returnOrigin} - ${returnDestination}`}</InfoText>
          <FactText>{returnFlightTime}</FactText>
        </TextBox>
        <TextBox>
          <InfoText>в пути</InfoText>
          <FactText>{returnFlightDuration}</FactText>
        </TextBox>
        <TextBox>
          <InfoText>
            {formatTransfersWordEnding(returnSegmentData.stops.length)}
          </InfoText>
          <FactText>{returnFlightTransfers}</FactText>
        </TextBox>
      </GroupBox>
    </CardBox>
  );
};

export default FlightCard;
