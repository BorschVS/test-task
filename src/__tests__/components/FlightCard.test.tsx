import '@testing-library/jest-dom';
import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithRouter } from 'utils/test-helpers/renderWithRouter';
import { renderWithProvider } from 'utils/test-helpers/renderWithProvider';

import * as actions from '../../redux/ducks/modalSlice';

import { FlightData } from 'types/interfaces';

import FlightCard from 'components/FlightCard';

const flightData: FlightData = {
  searchId: '11',
  id: '1',
  price: 10000,
  carrier: 'SU',
  segments: [
    {
      origin: 'City A',
      destination: 'City B',
      date: '2024-08-01T10:00:00Z',
      duration: 120,
      stops: ['SVA'],
    },
    {
      origin: 'City B',
      destination: 'City A',
      date: '2024-08-10T12:00:00Z',
      duration: 150,
      stops: ['SVA'],
    },
  ],
};

const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}));

describe('FlightCard Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('Should render flight details', async () => {
    renderWithRouter(<FlightCard flightData={flightData} />);

    await waitFor(() => {
      expect(screen.getByText(/10 000 Р/)).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.getByText(/10:00 - 12:00/)).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.getByText(/2Ч 0М/)).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.getAllByText(/1 пересадка/i).length).toBe(2);
    });

    await waitFor(() => {
      expect(screen.getByText(/City A - City B/i)).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.getByText(/City B - City A/i)).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.getAllByText(/SVA/)).toHaveLength(2);
    });
  });

  test('Should render modal on click', () => {
    const mockSetCurrentFlight = jest.spyOn(actions, 'setCurrentFlight');

    renderWithRouter(<FlightCard flightData={flightData} />);

    const flightCard = screen.getByTestId('flight-card');

    userEvent.click(flightCard);
    expect(mockDispatch).toHaveBeenCalled();
    expect(mockSetCurrentFlight).toHaveBeenCalledWith('1');
  });

  test('FlightCard snapshot', () => {
    const { asFragment } = renderWithProvider(
      <FlightCard flightData={flightData} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
