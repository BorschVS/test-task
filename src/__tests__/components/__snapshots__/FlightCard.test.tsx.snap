// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlightCard Component FlightCard snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-xv2kuy"
    data-testid="flight-card"
  >
    <div
      class="css-uf1ume"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1entie5-MuiTypography-root"
      >
        10 000 Р
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-g1npfy-MuiTypography-root"
      >
        <span>
          <span
            class="MuiAvatar-root MuiAvatar-circular css-106wf08-MuiAvatar-root"
          >
            <img
              alt="S7 Airlines company logo"
              class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
              src="/images/s7.jpg"
            />
          </span>
        </span>
        Airlines
      </p>
    </div>
    <div
      class="css-uf1ume"
    >
      <div
        class="css-1jibmi3"
      >
        <p
          class="css-9vg4wd"
        >
          City A - City B
        </p>
        <p
          class="css-6haxrs"
        >
          10:00 - 12:00
        </p>
      </div>
      <div
        class="css-1jibmi3"
      >
        <p
          class="css-9vg4wd"
        >
          в пути
        </p>
        <p
          class="css-6haxrs"
        >
          2Ч 0М
        </p>
      </div>
      <div
        class="css-1jibmi3"
      >
        <p
          class="css-9vg4wd"
        >
          1 пересадка
        </p>
        <p
          class="css-6haxrs"
        >
          SVA
        </p>
      </div>
    </div>
    <div
      class="css-uf1ume"
    >
      <div
        class="css-1jibmi3"
      >
        <p
          class="css-9vg4wd"
        >
          City B - City A
        </p>
        <p
          class="css-6haxrs"
        >
          12:00 - 14:30
        </p>
      </div>
      <div
        class="css-1jibmi3"
      >
        <p
          class="css-9vg4wd"
        >
          в пути
        </p>
        <p
          class="css-6haxrs"
        >
          2Ч 30М
        </p>
      </div>
      <div
        class="css-1jibmi3"
      >
        <p
          class="css-9vg4wd"
        >
          1 пересадка
        </p>
        <p
          class="css-6haxrs"
        >
          SVA
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
